unstructured_ingest/v2/processes/connectors/duckdb.py

```diff
@@ -0,0 +1,238 @@
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Callable, Literal, Optional
+
+import pandas as pd
+from pydantic import Field, Secret
+
+from unstructured_ingest.error import DestinationConnectionError
+from unstructured_ingest.utils.dep_check import requires_dependencies
+from unstructured_ingest.v2.interfaces import (
+    AccessConfig,
+    ConnectionConfig,
+    FileData,
+    Uploader,
+    UploaderConfig,
+    UploadStager,
+    UploadStagerConfig,
+)
+from unstructured_ingest.v2.logger import logger
+from unstructured_ingest.v2.processes.connector_registry import DestinationRegistryEntry
+
+if TYPE_CHECKING:
+    from duckdb import DuckDBPyConnection as DuckDBConnection
+
+CONNECTOR_TYPE = "duckdb"
+DUCKDB_DB = "duckdb"
+MOTHERDUCK_DB = "motherduck"
+
+
+class MotherDuckAccessConfig(AccessConfig):
+    md_token: Optional[str] = Field(default=None, description="MotherDuck token")
+
+
+class DuckDBConnectionConfig(ConnectionConfig):
+    db_type: Literal["duckdb", "motherduck"] = Field(
+        default=DUCKDB_DB, description="Type of the database backend"
+    )
+    database: Optional[str] = Field(
+        default=None,
+        description="Database name. This is the path to the DuckDB .db file.",
+    )
+    schema: Optional[str] = Field(
+        default="main",
+        description="Schema name. This is the schema within the database where the elements table is located.",
+    )
+    table: Optional[str] = Field(
+        default="elements",
+        description="Table name. This is the table name into which the elements data is inserted.",
+    )
+    access_config: Secret[MotherDuckAccessConfig] = Field(
+        default=MotherDuckAccessConfig(), validate_default=True
+    )
+    connector_type: str = Field(default=CONNECTOR_TYPE, init=False)
+
+    def __post_init__(self):
+        if (self.db_type == DUCKDB_DB) and (self.database is None):
+            raise ValueError(
+                "A DuckDB connection requires a path to a *.db or *.duckdb file "
+                "through the `database` argument"
+            )
+
+        if self.db_type == MOTHERDUCK_DB:
+            if self.database is None:
+                raise ValueError(
+                    "A MotherDuck connection requires a database (string) to be passed "
+                    "through the `database` argument"
+                )
+            if self.access_config.md_token is None:
+                raise ValueError(
+                    "A MotherDuck connection requires a md_token (MotherDuck token) to be passed "
+                    "using MotherDuckAccessConfig through the `access_config` argument"
+                )
+
+
+class DuckDBUploadStagerConfig(UploadStagerConfig):
+    pass
+
+
+_COLUMNS = (
+    "id",
+    "element_id",
+    "text",
+    "embeddings",
+    "type",
+    "system",
+    "layout_width",
+    "layout_height",
+    "points",
+    "url",
+    "version",
+    "date_created",
```

Comment:
These columns come in from partitioning as int64 and float64:


date_created                                             1728685124
date_modified                                            1728685124
date_processed                                    1728929563.441951


So the SQL insert is having trouble with them. Probably converting them to timestamp

Reply:
@potter-potter is this different from `last_modified` ? because when this line https://github.com/krish-adi/unstructured-ingest/blob/624371416d9661b1f5a81a2af0a2fa216fdd598b/unstructured_ingest/v2/processes/connectors/duckdb.py#L220 is executed, the df_elements stores the last_modified as this `2024-09-27T05:10:47`. 

is there a file or an option i can turn on to be able to test this out locally? the script i have to test locally never adds this metadata.

Reply:
Yeah. I can't remember why we get timestamps for last_modified and unix epoch numbers or whatever for the other ones.

Just try to intercept the DataFrame right before it writes and check out the first row.

Reply:
I will look to see if we even need them converted to timestamps, or if we just leave them as numbers when I'm back online tomorrow.

Reply:
this change: https://github.com/krish-adi/unstructured-ingest/blob/75774b7b5bcd9214302a77d0bc229d33c2bc75b3/test_e2e/env_setup/duckdb/create-duckdb-schema.sql#L13-L15

should hopefully resolve the insert conflicts.

Reply:
hey @potter-potter let me know if these changes are a possible workaround....